EIPS/eip-7620.md

```diff
@@ -0,0 +1,279 @@
+---
+eip: 7620
+title: EOF Contract Creation
+description: Introduce `CREATE3`, `CREATE4`, `RETURNCONTRACT` instructions along with a new `InitcodeTransaction` transaction
+author: Alex Beregszaszi (@axic), Paweł Bylica (@chfast), Andrei Maiboroda (@gumb0), Piotr Dobaczewski (@pdobacz)
+discussions-to: https://ethereum-magicians.org/t/eip-7620-eof-contract-creation-instructions/18625
+status: Draft
+type: Standards Track
+category: Core
+created: 2024-02-12
+requires: 170, 1559, 2028, 2718, 3540, 3541, 3670, 3860
+---
+
+## Abstract
+
+EVM Object Format (EOF) removes the possibility to create contracts using creation transactions (with an empty `to` field), `CREATE` or `CREATE2` instructions. We introduce three new instructions: `CREATE3`, `CREATE4` and `RETURNCONTRACT`, as well as a new transaction type (`InitcodeTransaction`) to provide a way to create contracts using EOF containers.
+
+## Motivation
+    
+This EIP uses terminology from the [EIP-3540](./eip-3540.md) which introduces the EOF format.
+
+EOF aims to remove code observability, which is a prerequisite to legacy EVM contract creation logic using create transactions, `CREATE` or `CREATE2`, because both the initcode and code are available to the EVM and can be manipulated. On the same premise, EOF removes opcodes like `CODECOPY` and `EXTCODECOPY`, introducing EOF subcontainers as a replacement to cater for factory contracts creating other contracts.
+    
+The new instructions and the new transaction type introduced in this EIP operate on EOF containers enabling all use cases of contract creation that legacy EVM has.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+    
+### Parameters
+
+| Constant | Value |
+| - | - |
+| `INITCODE_TX_TYPE` | `Bytes1(0x04)` |
+| `MAX_INITCODE_COUNT` | `256` |
+| `GKECCAK256WORD` | Defined as `6` in the Ethereum Yellow Paper |
+| `GCREATE` | Defined as `32000` in the Ethereum Yellow Paper |
+| `GCODEDEPOSIT` | Defined as `200` in the Ethereum Yellow Paper |
+| `GTXDATAZERO` | Defined as `4` in the Ethereum Yellow Paper |
+| `GTXDATANONZERO` | Defined as `16` in the Ethereum Yellow Paper |
+| `INITCODE_WORD_COST` | Defined as `2` in [EIP-3860](./eip-3860.md) |
+| `MAX_INITCODE_SIZE` | Defined as `2 * MAX_CODE_SIZE` in [EIP-3860](./eip-3860.md) |
+| `MAX_CODE_SIZE` | Defined as `24576` in [EIP-170](./eip-170.md) |
```

Comment:
As a suggestion, and not at all a hard requirement, you can link to the Yellow Paper or to EELS ([example](https://github.com/ethereum/execution-specs/blob/9ca885c8eb5e5151cda74f02e6675cd716764972/src/ethereum/shanghai/vm/interpreter.py#L61)) for these constants to provide more context.

Reply:
I have tried YP, that was rejected due to https://github.com/ethereum/EIPs/pull/8248. Didn't try EELS though, thanks!